### PR TITLE
Further S3 large file optimization

### LINKED ISSUE #7674: [DS-4334] S3 bitstore limits uploads to 5GB maximum (doesn't use Multipart upload)
<i>Imported from JIRA <a href="https://jira.lyrasis.org/browse/DS-4334">[DS-4334]</a> created by tdonohue</i>
<p>Bug logged per this discussion thread: <a href="https://groups.google.com/d/msg/dspace-tech/nLlSOzI3UZk/e_l4jhKUEAAJ" class="external-link" target="_blank" rel="nofollow noopener">https://groups.google.com/d/msg/dspace-tech/nLlSOzI3UZk/e_l4jhKUEAAJ</a></p>

<p>Simply put, the current implementation of S3BitStoreService does <b>not</b> use Amazon S3's multipart upload functionality in the "put" method: <a href="https://github.com/DSpace/DSpace/blob/dspace-6_x/dspace-api/src/main/java/org/dspace/storage/bitstore/S3BitStoreService.java#L164" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/DSpace/DSpace/blob/dspace-6_x/dspace-api/src/main/java/org/dspace/storage/bitstore/S3BitStoreService.java#L164</a></p>

<p>In order to use Multipart upload, this code would need to be updated to use a TransferManager, like the S3 example Java class here: <a href="https://docs.aws.amazon.com/AmazonS3/latest/dev/HLuploadFileJava.html" class="external-link" target="_blank" rel="nofollow noopener">https://docs.aws.amazon.com/AmazonS3/latest/dev/HLuploadFileJava.html</a></p>

<p><b>This means that currently the S3Bitstore only supports uploading files that are <=5GB</b>. This is because S3 only allows PUT commands of 5GB or less, unless you are using multipart upload: <a href="https://aws.amazon.com/s3/faqs/" class="external-link" target="_blank" rel="nofollow noopener">https://aws.amazon.com/s3/faqs/</a></p>

<p>Since S3 support files sizes of 5TB or more, you can still store / migrate larger files to S3 <b>manually</b>, but they cannot be uploaded via DSpace itself.</p>

<p> </p>

### LINKED ISSUE #8570: Integrity check & checksum report are broken (since v6)
**Describe the bug**
The `checker` script records all the check as success also when there are mismatch.
Moreover, the `checker-emailer` will never sent a report about identified issues.
 
**To Reproduce**
1. Manually remove some files from the assetstore and/or replace them with other content
2. run the `dspace checker` script and check the content of the checksum_results and checksum_history
3. run the `dspace checker-emailer` and verify the content of the received email

**Expected behavior**
The checksum_results table should contain the correct result for failing check (missing bitstreams or mismatching checksums).
The checker-emailer should report about recent failures

### PR DESCRIPTION
**Pull request by LucaGiamminonni**
## References
* Fixes https://github.com/DSpace/DSpace/issues/7674
* Fixes https://github.com/DSpace/DSpace/issues/8570

This PR added further S3 improvements in addition to those introduced with https://github.com/DSpace/DSpace/pull/8060

- [X] My PR is small in size (e.g. less than 1,000 lines of code, not including comments & integration tests). Exceptions may be made if previously agreed upon.
- [X] My PR passes Checkstyle validation based on the [Code Style Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Style+Guide).
- [X] My PR includes Javadoc for _all new (or modified) public methods and classes_. It also includes Javadoc for large or complex private methods.
- [X] My PR passes all tests and includes new/updated Unit or Integration Tests based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [X] If my PR includes new, third-party dependencies (in any `pom.xml`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/main/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.
- [X] If my PR modifies the REST API, I've linked to the REST Contract page (or open PR) related to this change.

### CODE REVIEW

**Comment by abollini** [comment]
@LucaGiamminonni I noted that test class has been heavily modified and I understand that the changes were due to the the tied coupling needed to mock the internal behavior of the S3 client inherited from the existing implementation. In this way the test are a self-referential and more intended to demonstrate that the mocking is working as expected than that proof the real code is working... is not a failure of this PR but would be possible to revisit the test to be more effective? would be possible to use a mock S3 embedded/service? Can you take a look at https://docs.tiledb.com/main/how-to/backends/minio and see if it is possible to use it for testing purpose?

File: dspace-api/pom.xml [hunk 1]
```diff
@@ -895,6 +913,11 @@
             <artifactId>swagger-core</artifactId>
             <version>1.6.2</version>
           </dependency>
+          <dependency>
+            <groupId>org.scala-lang</groupId>
+            <artifactId>scala-library</artifactId>
+            <version>2.13.2</version>
```
Comment (tdonohue) [inline dspace-api/pom.xml:919]: Should this have a `<scope>test</scope>` added?  It appears, to me, that we are not using this library in our code, but maybe it's required by the new `s3mock_2.13` library?  If so, we should make sure this is also limited to `test` scope.

File: dspace-api/src/main/java/org/dspace/storage/bitstore/S3BitStoreService.java [hunk 2]
```diff
@@ -325,8 +345,28 @@ public Map about(Bitstream bitstream, Map attrs) throws IOException {
         try {
             ObjectMetadata objectMetadata = s3Service.getObjectMetadata(bucketName, key);
             if (objectMetadata != null) {
-                return this.about(objectMetadata, attrs);
+                if (attrs.containsKey("size_bytes")) {
+                    attrs.put("size_bytes", objectMetadata.getContentLength());
```
Comment (tdonohue) [inline dspace-api/src/main/java/org/dspace/storage/bitstore/S3BitStoreService.java:349]: Could we add tests for the changes to this `about()` method?  Codecov is saying that none of the newly added code in this method is tested.

**Review by tdonohue — CHANGES_REQUESTED** [review]
@LucaGiamminonni : I gave this a code review today.  Overall the code looks good, but I have a few minor comments inline which I think should be fixed before this can be moved forward.

Unfortunately, I'm not able to test this myself as I don't have a testable S3 setup.  But, I'm hoping that @benbosman (or someone on his team) could find some time to give this a quick test.  Based on the integration tests though, this appears like it *should* work.

File: dspace-api/src/main/java/org/dspace/storage/bitstore/S3BitStoreService.java [hunk 3]
```diff
@@ -277,24 +291,30 @@
...
-        try {
-            FileUtils.copyInputStreamToFile(in, scratchFile);
-            long contentLength = scratchFile.length();
-            // The ETag may or may not be and MD5 digest of the object data.
-            // Therefore, we precalculate before uploading
-            String localChecksum = org.dspace.curate.Utils.checksum(scratchFile, CSA);
+        try (
+                FileOutputStream fos = new FileOutputStream(scratchFile);
+                // Read through a digest input stream that will work out the MD5
+                DigestInputStream dis = new DigestInputStream(in, MessageDigest.getInstance(CSA));
```
[the remark this answers is not in the document]
Comment (tdonohue) [reply dspace-api/src/main/java/org/dspace/storage/bitstore/S3BitStoreService.java:297]: This is a false positive.  MD5 is only being used for checksums, not for encryption.

File: dspace-api/src/main/java/org/dspace/storage/bitstore/S3BitStoreService.java [hunk 4]
```diff
@@ -309,61 +329,53 @@
...
-                return this.about(objectMetadata, attrs);
+                putValueIfExistsKey(attrs, metadata, "size_bytes", objectMetadata.getContentLength());
+                putValueIfExistsKey(attrs, metadata, "modified", valueOf(objectMetadata.getLastModified().getTime()));
             }
+
+            putValueIfExistsKey(attrs, metadata, "checksum_algorithm", CSA);
+
+            if (attrs.contains("checksum")) {
+                try (InputStream in = get(bitstream)) {
+                    byte[] md5Digest = MessageDigest.getInstance(CSA).digest(IOUtils.toByteArray(in));
```
[the remark this answers is not in the document]
Comment (tdonohue) [reply dspace-api/src/main/java/org/dspace/storage/bitstore/S3BitStoreService.java:361]: Same here. This is a false positive.  MD5 is only being used for checksums, not for encryption.

**Review by tdonohue — COMMENTED** [review]
@LucaGiamminonni : Gave this another review today.  The code itself looks good & the tests look good.  I'm not able to test the new S3 code, but I tried testing the fix for #8570.  Unfortunately, though, I'm not able to get a status other than CHECKSUM_MATCH in the `checksum_history` table.

I feel like I must be doing something wrong, but here's what I've tried:
1. First, I found a bitstream's assetstore location (`internal_id`) by running `select internal_id from bitstream where uuid='[bitstream-uuid]` in the database;
2. I went to the corresponding directory in the `[dspace]/assetstore/` and I created a "tmp" subdirectory and **moved** the bitstream file to that subdirectory.  (This should mean the bitstream is now "missing" as it's not in the correct directory)
3. I ran `./dspace checker -b [bitstream-uuid]`.
4. Back in the database I ran `select * from checksum_history where bitstream_id='[bitstream-uuid]';`  I found the most recent run, but it still says "CHECKSUM_MATCH".  (I thought it should now say "BITSTREAM_NOT_FOUND").

Overall, the code here looks good & looks like an improvement.  But, I'm not sure this fully fixes #8570, unless I'm misunderstanding something.

**Comment by toniprieto** [comment]
@LucaGiamminonni @tdonohue About the checksum checker issue, in case it helps, there is a 6x PR that solve the same problem: https://github.com/DSpace/DSpace/pull/2160

It has an extra modification on [dspace-api/src/main/java/org/dspace/checker/CheckerCommand.java](https://github.com/DSpace/DSpace/pull/2160/files#diff-0e20910334231dc6e037c25d550acfbdbacdd287a6c4800a913522ba9fcfb1e7) that could solve the issue reported by Tim in the last comment.

Moreover, it has another modification similar to https://github.com/DSpace/DSpace/pull/8500/files#diff-ac4d9e325a8d736647418f4d16e4f7721cc0564eaa4565b0acb29b9b27a0764f for function `findByNotProcessedInDateRange`

I tested in a 6x instance and it seems to work ok.

**Review by benbosman — COMMENTED** [review]
I've also reviewed the code, and don't see any issues.

I will try to test the functionality with S3 as well

Thanks for the contribution

**Comment by tdonohue** [comment]
Moving this to 7.6 as there's been no updates in the last month & there's feedback above to be resolved..  We've run out of time for 7.5, but I'd love to see this move forward for 7.6.

**Review by tdonohue — APPROVED** [review]
👍 Thanks @LucaGiamminonni !  I've verified this fixes the issues with checksum history now (described in #8570).  The code for S3 also looks good and the tests look good as well.

I'm not able to test this against a DSpace running in S3, so I'd appreciate it if @benbosman (or someone on his team) could run a quick test prior to merger.  But, the code looks good to me.

**Comment by abollini** [comment]
hi @tdonohue I checked and our internal code that we use for hosted DSpace and the code in DSpace-CRIS is slightly different.
Mainly because in this PR we have further cleanup the code but we also have some additional capacities on our code that was introduced later and not included in this PR. So a real test would be better, @francescopioscognamiglio  will try to do tomorrow or early next week at latest

**Comment by tdonohue** [comment]
@abollini : Thank you to you and @francescopioscognamiglio !  Once a test is complete, just comment back here to let me know how it goes.  I simply need verification that the updated S3 code in this PR is working well, as I have no way to easily test it myself.

**Comment by francescopioscognamiglio** [comment]
Hi @tdonohue I have done the following tests using an S3 assetstore.
**Test of existing bitstreams:**
1. open an item with at least one bitstream
2. download the bitstream

**Test of a new bitstream:**
1. start a new submission
2. upload a bitstream
3. finish the submission
4. download the bitstream

**Comment by tdonohue** [comment]
@francescopioscognamiglio : Thank you for the testing!  In that case, I will merge this for 7.6, as the code looks good to me & I previously verified it fixes the checksum checker bug... and just wanted verification that S3 was working.
